### DISCUSSION TITLE
Determine when using cached response (onSuccess is not called when cache is used)

### DISCUSSION THREAD

**Discussion by calengbuford** [TanStack/query#3335] [Ideas]
When successfully fetching a new response, the onSuccess function is called. However, when using the cached response instead, onSuccess is not called. This means if I want to send analytics (or do anything else) upon a successful data change, I can't do that with onSuccess.

It would be great to have a new function that tells me when the cache is used or that tells me when the data is updated. A useEffect to manually do this can be costly for performance.

**Comment by TkDodo** [comment]
> A useEffect to manually do this can be costly for performance.

why? A `useEffect` with `data` as dependency does exactly what you want - being executed whenever `data` changes.

For v4, we thought about an extra callback `onDataChanged`, but it's not a very common use-case and can easily be done with an effect in userland. Also [documented](https://react-query-alpha.tanstack.com/guides/migrating-to-react-query-4#onsuccess-is-no-longer-called-from-setquerydata) this in v4.
